Flaky test TestGoogleSheets

```
2020-02-25T07:42:15.7112549Z [ERROR] testSelectFromTable(io.prestosql.plugin.google.sheets.TestGoogleSheets)  Time elapsed: 16.317 s  <<< FAILURE!
2020-02-25T07:42:15.7119018Z java.lang.AssertionError: Execution of 'actual' query failed: SELECT count(*) FROM number_text
2020-02-25T07:42:15.7122354Z 	at org.testng.Assert.fail(Assert.java:83)
2020-02-25T07:42:15.7127844Z 	at io.prestosql.testing.QueryAssertions.assertQuery(QueryAssertions.java:147)
2020-02-25T07:42:15.7133104Z 	at io.prestosql.testing.QueryAssertions.assertQuery(QueryAssertions.java:103)
2020-02-25T07:42:15.7139512Z 	at io.prestosql.testing.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:130)
2020-02-25T07:42:15.7145919Z 	at io.prestosql.plugin.google.sheets.TestGoogleSheets.testSelectFromTable(TestGoogleSheets.java:74)
2020-02-25T07:42:15.7150330Z 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2020-02-25T07:42:15.7156701Z 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2020-02-25T07:42:15.7163148Z 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2020-02-25T07:42:15.7167300Z 	at java.lang.reflect.Method.invoke(Method.java:498)
2020-02-25T07:42:15.7175282Z 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
2020-02-25T07:42:15.7179682Z 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
2020-02-25T07:42:15.7184270Z 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
2020-02-25T07:42:15.7188927Z 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
2020-02-25T07:42:15.7194553Z 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
2020-02-25T07:42:15.7199400Z 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
2020-02-25T07:42:15.7204857Z 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
2020-02-25T07:42:15.7210543Z 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
2020-02-25T07:42:15.7214034Z 	at java.lang.Thread.run(Thread.java:748)
2020-02-25T07:42:15.7222091Z Caused by: java.lang.RuntimeException: Failed reading data from sheet: 1Es4HhWALUQjoa-bQh4a8B5HROz7dpGMfq_HbfoaW5LM#Tables
2020-02-25T07:42:15.7228441Z 	at io.prestosql.testing.AbstractTestingPrestoClient.execute(AbstractTestingPrestoClient.java:114)
2020-02-25T07:42:15.7234214Z 	at io.prestosql.testing.DistributedQueryRunner.execute(DistributedQueryRunner.java:400)
2020-02-25T07:42:15.7240170Z 	at io.prestosql.testing.QueryAssertions.assertQuery(QueryAssertions.java:144)
2020-02-25T07:42:15.7241375Z 	... 16 more
2020-02-25T07:42:15.7252340Z Caused by: io.prestosql.spi.PrestoException: Failed reading data from sheet: 1Es4HhWALUQjoa-bQh4a8B5HROz7dpGMfq_HbfoaW5LM#Tables
2020-02-25T07:42:15.7258736Z 	at io.prestosql.plugin.google.sheets.SheetsClient.readAllValuesFromSheetExpression(SheetsClient.java:216)
2020-02-25T07:42:15.7265357Z 	at io.prestosql.plugin.google.sheets.SheetsClient.getAllTableSheetExpressionMapping(SheetsClient.java:180)
2020-02-25T07:42:15.7271760Z 	at io.prestosql.plugin.google.sheets.SheetsClient.getSheetExpressionForTable(SheetsClient.java:170)
2020-02-25T07:42:15.7277185Z 	at io.prestosql.plugin.google.sheets.SheetsClient.access$000(SheetsClient.java:58)
2020-02-25T07:42:15.7282399Z 	at io.prestosql.plugin.google.sheets.SheetsClient$1.load(SheetsClient.java:98)
2020-02-25T07:42:15.7287829Z 	at io.prestosql.plugin.google.sheets.SheetsClient$1.load(SheetsClient.java:94)
2020-02-25T07:42:15.7293799Z 	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3528)
2020-02-25T07:42:15.7298922Z 	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2277)
2020-02-25T07:42:15.7304391Z 	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2154)
2020-02-25T07:42:15.7309216Z 	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2044)
2020-02-25T07:42:15.7313616Z 	at com.google.common.cache.LocalCache.get(LocalCache.java:3952)
2020-02-25T07:42:15.7319113Z 	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3974)
2020-02-25T07:42:15.7324511Z 	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4958)
2020-02-25T07:42:15.7330974Z 	at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4964)
2020-02-25T07:42:15.7336711Z 	at io.prestosql.plugin.google.sheets.SheetsClient.readAllValues(SheetsClient.java:156)
2020-02-25T07:42:15.7342114Z 	at io.prestosql.plugin.google.sheets.SheetsClient.getTable(SheetsClient.java:113)
2020-02-25T07:42:15.7347958Z 	at io.prestosql.plugin.google.sheets.SheetsMetadata.getTableHandle(SheetsMetadata.java:73)
2020-02-25T07:42:15.7353939Z 	at io.prestosql.plugin.google.sheets.SheetsMetadata.getTableHandle(SheetsMetadata.java:42)
2020-02-25T07:42:15.7359315Z 	at io.prestosql.metadata.MetadataManager.getTableHandle(MetadataManager.java:312)
2020-02-25T07:42:15.7365847Z 	at io.prestosql.sql.rewrite.ShowQueriesRewrite$Visitor.visitShowColumns(ShowQueriesRewrite.java:359)
2020-02-25T07:42:15.7372233Z 	at io.prestosql.sql.rewrite.ShowQueriesRewrite$Visitor.visitShowColumns(ShowQueriesRewrite.java:153)
2020-02-25T07:42:15.7376685Z 	at io.prestosql.sql.tree.ShowColumns.accept(ShowColumns.java:54)
2020-02-25T07:42:15.7381076Z 	at io.prestosql.sql.tree.AstVisitor.process(AstVisitor.java:27)
2020-02-25T07:42:15.7386541Z 	at io.prestosql.sql.rewrite.ShowQueriesRewrite.rewrite(ShowQueriesRewrite.java:150)
2020-02-25T07:42:15.7391744Z 	at io.prestosql.sql.rewrite.StatementRewrite.rewrite(StatementRewrite.java:57)
2020-02-25T07:42:15.7396140Z 	at io.prestosql.sql.analyzer.Analyzer.analyze(Analyzer.java:80)
2020-02-25T07:42:15.7486697Z 	at io.prestosql.sql.analyzer.Analyzer.analyze(Analyzer.java:75)
2020-02-25T07:42:15.7488574Z 	at io.prestosql.execution.SqlQueryExecution.analyze(SqlQueryExecution.java:221)
2020-02-25T07:42:15.7491033Z 	at io.prestosql.execution.SqlQueryExecution.<init>(SqlQueryExecution.java:180)
2020-02-25T07:42:15.7493377Z 	at io.prestosql.execution.SqlQueryExecution.<init>(SqlQueryExecution.java:97)
2020-02-25T07:42:15.7496393Z 	at io.prestosql.execution.SqlQueryExecution$SqlQueryExecutionFactory.createQueryExecution(SqlQueryExecution.java:732)
2020-02-25T07:42:15.7499538Z 	at io.prestosql.dispatcher.LocalDispatchQueryFactory.lambda$createDispatchQuery$0(LocalDispatchQueryFactory.java:121)
2020-02-25T07:42:15.7502376Z 	at io.prestosql.$gen.Presto_330_81_gad394cb__testversion____20200225_074137_3.call(Unknown Source)
2020-02-25T07:42:15.7506067Z 	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
2020-02-25T07:42:15.7508679Z 	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
2020-02-25T07:42:15.7511542Z 	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
2020-02-25T07:42:15.7512856Z 	... 3 more
2020-02-25T07:42:15.7514658Z Caused by: java.net.SocketTimeoutException: Read timed out
2020-02-25T07:42:15.7516617Z 	at java.net.SocketInputStream.socketRead0(Native Method)
2020-02-25T07:42:15.7518785Z 	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
2020-02-25T07:42:15.7520833Z 	at java.net.SocketInputStream.read(SocketInputStream.java:171)
2020-02-25T07:42:15.7522805Z 	at java.net.SocketInputStream.read(SocketInputStream.java:141)
2020-02-25T07:42:15.7524798Z 	at sun.security.ssl.InputRecord.readFully(InputRecord.java:465)
2020-02-25T07:42:15.7526977Z 	at sun.security.ssl.InputRecord.read(InputRecord.java:503)
2020-02-25T07:42:15.7529120Z 	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:975)
2020-02-25T07:42:15.7531382Z 	at sun.security.ssl.SSLSocketImpl.readDataRecord(SSLSocketImpl.java:933)
2020-02-25T07:42:15.7548764Z 	at sun.security.ssl.AppInputStream.read(AppInputStream.java:105)
2020-02-25T07:42:15.7550690Z 	at java.io.BufferedInputStream.fill(BufferedInputStream.java:246)
2020-02-25T07:42:15.7552704Z 	at java.io.BufferedInputStream.read1(BufferedInputStream.java:286)
2020-02-25T07:42:15.7554801Z 	at java.io.BufferedInputStream.read(BufferedInputStream.java:345)
2020-02-25T07:42:15.7556966Z 	at sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:743)
2020-02-25T07:42:15.7558914Z 	at sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:678)
2020-02-25T07:42:15.7561391Z 	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1593)
2020-02-25T07:42:15.7563910Z 	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1498)
2020-02-25T07:42:15.7566412Z 	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480)
2020-02-25T07:42:15.7569095Z 	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:352)
2020-02-25T07:42:15.7571582Z 	at com.google.api.client.http.javanet.NetHttpResponse.<init>(NetHttpResponse.java:37)
2020-02-25T07:42:15.7573936Z 	at com.google.api.client.http.javanet.NetHttpRequest.execute(NetHttpRequest.java:105)
2020-02-25T07:42:15.7576138Z 	at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:981)
2020-02-25T07:42:15.7579354Z 	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:419)
2020-02-25T07:42:15.7582466Z 	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:352)
2020-02-25T07:42:15.7585471Z 	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:469)
2020-02-25T07:42:15.7588254Z 	at io.prestosql.plugin.google.sheets.SheetsClient.readAllValuesFromSheetExpression(SheetsClient.java:213)
2020-02-25T07:42:15.7589572Z 	... 38 more
```

--- a/docs/src/main/sphinx/connector/googlesheets.rst
+++ b/docs/src/main/sphinx/connector/googlesheets.rst
@@ -33,6 +33,7 @@ Property name                       Description
 ``gsheets.metadata-sheet-id``       Sheet ID of the spreadsheet, that contains the table mapping
 ``gsheets.max-data-cache-size``     Maximum number of spreadsheets to cache, defaults to ``1000``
 ``gsheets.data-cache-ttl``          How long to cache spreadsheet data or metadata, defaults to ``5m``
+``gsheets.read-timeout``            Timeout to read data from spreadsheet, defaults to ``20s``
 =================================== =====================================================================
 
 Credentials

--- a/plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsClient.java
+++ b/plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsClient.java
@@ -15,6 +15,7 @@ package io.trino.plugin.google.sheets;
 
 import com.google.api.client.auth.oauth2.Credential;
 import com.google.api.client.googleapis.auth.oauth2.GoogleCredential;
+import com.google.api.client.http.HttpRequestInitializer;
 import com.google.api.client.json.JsonFactory;
 import com.google.api.client.json.jackson2.JacksonFactory;
 import com.google.api.services.sheets.v4.Sheets;
@@ -27,6 +28,7 @@ import com.google.common.collect.ImmutableSet;
 import com.google.common.util.concurrent.UncheckedExecutionException;
 import io.airlift.json.JsonCodec;
 import io.airlift.log.Logger;
+import io.airlift.units.Duration;
 import io.trino.collect.cache.NonEvictableLoadingCache;
 import io.trino.spi.TrinoException;
 import io.trino.spi.type.VarcharType;
@@ -52,6 +54,7 @@ import static io.trino.plugin.google.sheets.SheetsErrorCode.SHEETS_BAD_CREDENTIA
 import static io.trino.plugin.google.sheets.SheetsErrorCode.SHEETS_METASTORE_ERROR;
 import static io.trino.plugin.google.sheets.SheetsErrorCode.SHEETS_TABLE_LOAD_ERROR;
 import static io.trino.plugin.google.sheets.SheetsErrorCode.SHEETS_UNKNOWN_TABLE_ERROR;
+import static java.lang.Math.toIntExact;
 import static java.util.Locale.ENGLISH;
 import static java.util.Objects.requireNonNull;
 import static java.util.concurrent.TimeUnit.MILLISECONDS;
@@ -82,7 +85,7 @@ public class SheetsClient
         this.credentialsFilePath = config.getCredentialsFilePath();
 
         try {
-            this.sheetsService = new Sheets.Builder(newTrustedTransport(), JSON_FACTORY, getCredentials()).setApplicationName(APPLICATION_NAME).build();
+            this.sheetsService = new Sheets.Builder(newTrustedTransport(), JSON_FACTORY, setTimeout(getCredentials(), config.getReadTimeout())).setApplicationName(APPLICATION_NAME).build();
         }
         catch (GeneralSecurityException | IOException e) {
             throw new TrinoException(SHEETS_BAD_CREDENTIALS_ERROR, e);
@@ -229,5 +232,14 @@ public class SheetsClient
     private static CacheBuilder<Object, Object> newCacheBuilder(long expiresAfterWriteMillis, long maximumSize)
     {
         return CacheBuilder.newBuilder().expireAfterWrite(expiresAfterWriteMillis, MILLISECONDS).maximumSize(maximumSize);
+    }
+
+    private static HttpRequestInitializer setTimeout(HttpRequestInitializer requestInitializer, Duration readTimeout)
+    {
+        requireNonNull(readTimeout, "readTimeout is null");
+        return httpRequest -> {
+            requestInitializer.initialize(httpRequest);
+            httpRequest.setReadTimeout(toIntExact(readTimeout.toMillis()));
+        };
     }
 }

--- a/plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsConfig.java
+++ b/plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsConfig.java
@@ -31,6 +31,7 @@ public class SheetsConfig
     private String metadataSheetId;
     private int sheetsDataMaxCacheSize = 1000;
     private Duration sheetsDataExpireAfterWrite = new Duration(5, TimeUnit.MINUTES);
+    private Duration readTimeout = new Duration(20, TimeUnit.SECONDS); // 20s is the default timeout of com.google.api.client.http.HttpRequest
 
     @NotNull
     @FileExists
@@ -90,6 +91,19 @@ public class SheetsConfig
     public SheetsConfig setSheetsDataExpireAfterWrite(Duration sheetsDataExpireAfterWriteMinutes)
     {
         this.sheetsDataExpireAfterWrite = sheetsDataExpireAfterWriteMinutes;
+        return this;
+    }
+
+    @MinDuration("0ms")
+    public Duration getReadTimeout()
+    {
+        return readTimeout;
+    }
+
+    @Config("gsheets.read-timeout")
+    public SheetsConfig setReadTimeout(Duration readTimeout)
+    {
+        this.readTimeout = readTimeout;
         return this;
     }
 }

--- a/plugin/trino-google-sheets/src/test/java/io/trino/plugin/google/sheets/TestGoogleSheets.java
+++ b/plugin/trino-google-sheets/src/test/java/io/trino/plugin/google/sheets/TestGoogleSheets.java
@@ -28,7 +28,7 @@ public class TestGoogleSheets
     protected QueryRunner createQueryRunner()
             throws Exception
     {
-        return createSheetsQueryRunner(ImmutableMap.of(), ImmutableMap.of());
+        return createSheetsQueryRunner(ImmutableMap.of(), ImmutableMap.of("gsheets.read-timeout", "1m"));
     }
 
     @Test

--- a/plugin/trino-google-sheets/src/test/java/io/trino/plugin/google/sheets/TestSheetsConfig.java
+++ b/plugin/trino-google-sheets/src/test/java/io/trino/plugin/google/sheets/TestSheetsConfig.java
@@ -37,7 +37,8 @@ public class TestSheetsConfig
                 .setCredentialsFilePath(null)
                 .setMetadataSheetId(null)
                 .setSheetsDataMaxCacheSize(1000)
-                .setSheetsDataExpireAfterWrite(new Duration(5, TimeUnit.MINUTES)));
+                .setSheetsDataExpireAfterWrite(new Duration(5, TimeUnit.MINUTES))
+                .setReadTimeout(new Duration(20, TimeUnit.SECONDS)));
     }
 
     @Test
@@ -51,13 +52,15 @@ public class TestSheetsConfig
                 .put("gsheets.metadata-sheet-id", "foo_bar_sheet_id#Sheet1")
                 .put("gsheets.max-data-cache-size", "2000")
                 .put("gsheets.data-cache-ttl", "10m")
+                .put("gsheets.read-timeout", "1m")
                 .buildOrThrow();
 
         SheetsConfig expected = new SheetsConfig()
                 .setCredentialsFilePath(credentialsFile.toString())
                 .setMetadataSheetId("foo_bar_sheet_id#Sheet1")
                 .setSheetsDataMaxCacheSize(2000)
-                .setSheetsDataExpireAfterWrite(new Duration(10, TimeUnit.MINUTES));
+                .setSheetsDataExpireAfterWrite(new Duration(10, TimeUnit.MINUTES))
+                .setReadTimeout(new Duration(1, TimeUnit.MINUTES));
 
         assertFullMapping(properties, expected);
     }

